server/src/main/java/org/opensearch/search/backpressure/SearchBackpressureService.java

```diff
@@ -0,0 +1,290 @@
+
+    private final List<NodeResourceUsageTracker> nodeResourceUsageTrackers;
+    private final List<TaskResourceUsageTracker> taskResourceUsageTrackers;
+
+    private final AtomicReference<TokenBucket> taskCancellationRateLimiter = new AtomicReference<>();
+    private final AtomicReference<TokenBucket> taskCancellationRatioLimiter = new AtomicReference<>();
+
+    // Currently, only the state of SearchShardTask is being tracked.
+    // This can be generalized to Map<TaskType, SearchBackpressureState> once we start supporting cancellation of SearchTasks as well.
+    private final SearchBackpressureState state = new SearchBackpressureState();
+
+    public SearchBackpressureService(
+        SearchBackpressureSettings settings,
+        TaskResourceTrackingService taskResourceTrackingService,
+        ThreadPool threadPool
+    ) {
+        this(
+            settings,
+            taskResourceTrackingService,
+            threadPool,
+            System::nanoTime,
+            List.of(
+                new NodeResourceUsageTracker(
+                    () -> ProcessProbe.getInstance().getProcessCpuPercent() / 100.0 >= settings.getNodeDuressCpuThreshold()
+                ),
+                new NodeResourceUsageTracker(
+                    () -> JvmStats.jvmStats().getMem().getHeapUsedPercent() / 100.0 >= settings.getNodeDuressHeapThreshold()
+                )
+            ),
+            Collections.emptyList()
+        );
+    }
+
+    public SearchBackpressureService(
+        SearchBackpressureSettings settings,
+        TaskResourceTrackingService taskResourceTrackingService,
+        ThreadPool threadPool,
+        LongSupplier timeNanosSupplier,
+        List<NodeResourceUsageTracker> nodeResourceUsageTrackers,
+        List<TaskResourceUsageTracker> taskResourceUsageTrackers
+    ) {
+        this.settings = settings;
+        this.settings.setListener(this);
+        this.taskResourceTrackingService = taskResourceTrackingService;
+        this.taskResourceTrackingService.addTaskCompletionListener(this);
+        this.timeNanosSupplier = timeNanosSupplier;
+        this.nodeResourceUsageTrackers = nodeResourceUsageTrackers;
+        this.taskResourceUsageTrackers = taskResourceUsageTrackers;
+
+        this.taskCancellationRateLimiter.set(
+            new TokenBucket(timeNanosSupplier, getSettings().getCancellationRateNanos(), getSettings().getCancellationBurst())
+        );
+
+        this.taskCancellationRatioLimiter.set(
+            new TokenBucket(state::getCompletionCount, getSettings().getCancellationRatio(), getSettings().getCancellationBurst())
+        );
+
+        threadPool.scheduleWithFixedDelay(this, getSettings().getInterval(), ThreadPool.Names.GENERIC);
+    }
+
+    @Override
+    public void run() {
+        try {
+            doRun();
+        } catch (Exception e) {
+            logger.debug("failure in search search backpressure", e);
+        }
+    }
+
+    public void doRun() {
```

Comment:
Is `public` intentional so that the same logic can be run across the lifetime of the search task

Reply:
Updated it to package private (needed for tests).